File: src/commands/videos.ts
import { getBaseDir, getOutputFormat, shouldSilenceLogs, sanitizeFilename } from "../lib/utils.ts";
import { Command } from "commander";
import type { Logger, SessionInfo, OutputFormat } from "../lib/types.ts";
import { getEnrolledCourses, getEnrolledCoursesApi, getSupervideosInCourse, getSupervideosInCourseApi, getVideoMetadata, completeVideoApi, completeVideo, downloadVideo, getIncompleteVideosApi } from "../lib/moodle.ts";
import { createLogger } from "../lib/logger.ts";
import { launchAuthenticated } from "../lib/auth.ts";
import { extractSessionInfo } from "../lib/session.ts";
import { closeBrowserSafely } from "../lib/auth.ts";
import { formatAndOutput } from "../index.ts";
import { loadWsToken } from "../lib/token.ts";
import path from "node:path";
import fs from "node:fs";

export function registerVideosCommand(program: Command): void {
  const videosCmd = program.command("videos");
  videosCmd.description("Video progress operations");

  // Pure API context - no browser required (fast!)
  async function createApiContext(options: { verbose?: boolean; headed?: boolean }, command?: any): Promise<{
    log: Logger;
    session: { wsToken: string; moodleBaseUrl: string };
  } | null> {
    const opts = command?.optsWithGlobals ? command.optsWithGlobals() : options;
    // Don't silence logs for commands that don't have explicit output format control
    const outputFormat = command && command.optsWithGlobals ? getOutputFormat(command) : "table";
    const silent = outputFormat === "json" && !opts.verbose;
    const log = createLogger(opts.verbose, silent);

    const baseDir = getBaseDir();
    const sessionPath = path.resolve(baseDir, ".auth", "storage-state.json");

    // Check if session exists
    if (!fs.existsSync(sessionPath)) {
      log.error("未找到登入 session。請先執行 'openape auth login' 進行登入。");
      log.info(`Session 預期位置: ${sessionPath}`);
      return null;
    }

    // Try to load WS token
    const wsToken = loadWsToken(sessionPath);
    if (!wsToken) {
      log.error("未找到 WS token。請先執行 'openape auth login' 進行登入。");
      return null;
    }

    return {
      log,
      session: {
        wsToken,
        moodleBaseUrl: "https://ilearning.cycu.edu.tw",
      },
    };
  }

  // Helper function to create session context (for browser-only commands)
  async function createSessionContext(options: { verbose?: boolean; headed?: boolean }, command?: any): Promise<{
    log: Logger;
    page: import("playwright-core").Page;
    session: SessionInfo;
    browser: any;
    context: any;
  } | null> {
    const opts = command?.optsWithGlobals ? command.optsWithGlobals() : options;
    const outputFormat = getOutputFormat(command || { optsWithGlobals: () => ({ output: "json" }) });
    const silent = outputFormat === "json" && !opts.verbose;
    const log = createLogger(opts.verbose, silent);

    const baseDir = getBaseDir();
    const sessionPath = path.resolve(baseDir, ".auth", "storage-state.json");

    if (!fs.existsSync(sessionPath)) {
      log.error("未找到登入 session。請先執行 'openape auth login' 進行登入。");
      return null;
    }

    const config = {
      username: "",
      password: "",
      courseUrl: "",
      moodleBaseUrl: "https://ilearning.cycu.edu.tw",
      headless: !options.headed,
      slowMo: 0,
      authStatePath: sessionPath,
      ollamaBaseUrl: "",
    };

    log.info("啟動瀏覽器...");
    const { browser, context, page } = await launchAuthenticated(config, log);

    try {
      const session = await extractSessionInfo(page, config, log);
      return { log, page, session, browser, context };
    } catch (err) {
      await context.close();
      await browser.close();
      throw err;
    }
  }

  videosCmd
    .command("list")
    .description("List videos in a course")
    .argument("<course-id>", "Course ID")
    .option("--incomplete-only", "Show only incomplete videos")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      let videos = await getSupervideosInCourseApi(apiContext.session, parseInt(courseId, 10));

      // Filter for incomplete videos if requested
      if (options.incompleteOnly) {
        videos = videos.filter(v => !v.isComplete);
      }

      formatAndOutput(videos as unknown as Record<string, unknown>[], output, apiContext.log);
    });

  videosCmd
    .command("complete")
    .description("Complete videos in a course (uses API for list & completion, browser for metadata)")
    .argument("<course-id>", "Course ID")
    .option("--dry-run", "Discover videos but don't complete them")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (courseId, options, command) => {
      const output: OutputFormat = getOutputFormat(command);

      // Get API context for getting incomplete videos and completion
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      // Get incomplete videos via API (fast, no browser needed)
      const incompleteVideos = await getIncompleteVideosApi(apiContext.session, parseInt(courseId, 10));

      if (incompleteVideos.length === 0) {
        apiContext.log.info("所有影片已完成（或無影片）。");
        return;
      }

      apiContext.log.info(`找到 ${incompleteVideos.length} 部未完成影片`);

      // Dry-run: show videos without needing browser
      if (options.dryRun) {
        const results = incompleteVideos.map(v => ({ name: v.name, success: true }));
        for (const video of incompleteVideos) {
          apiContext.log.info(`  [試執行] ${video.name}`);
        }
        apiContext.log.info(`\n執行結果: ${results.length} 影片將被完成`);

        if (output !== "silent") {
          formatAndOutput(results as unknown as Record<string, unknown>[], output, apiContext.log);
        }
        return;
      }

      // Need browser only for getting viewId and duration (not needed for dry-run)
      const context = await createSessionContext(options, command);
      if (!context) {
        process.exitCode = 1;
        return;
      }

      const { log, page, browser, context: browserContext } = context;

      try {
        const results: Array<{ name: string; success: boolean; error?: string }> = [];

        for (const sv of incompleteVideos) {
          log.info(`處理中: ${sv.name}`);

          try {
            const video = await getVideoMetadata(page, sv.url, log);

            // Use WS API for completion
            const success = await completeVideoApi(apiContext.session, { ...video, cmid: sv.cmid.toString() });
            if (success) {
              log.success(`  已完成！`);
              results.push({ name: sv.name, success: true });
            } else {
              log.error(`  失敗。`);
              results.push({ name: sv.name, success: false, error: "Failed to complete" });
            }
          } catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            log.error(`  錯誤: ${msg}`);
            results.push({ name: sv.name, success: false, error: msg });
          }
        }

        const completed = results.filter(r => r.success).length;
        const failed = results.filter(r => !r.success).length;
        log.info(`\n執行結果: ${completed} 成功, ${failed} 失敗`);

        if (output !== "silent") {
          formatAndOutput(results as unknown as Record<string, unknown>[], output, log);
        }
      } finally {
        await closeBrowserSafely(browser, browserContext);
      }
    });

  videosCmd
    .command("complete-all")
    .description("Complete all incomplete videos across all courses (uses API for list & completion, browser for metadata)")
    .option("--dry-run", "Discover videos but don't complete them")
    .option("--output <format>", "Output format: json|csv|table|silent")
    .action(async (options, command) => {
      const output: OutputFormat = getOutputFormat(command);

      // Get API context for getting incomplete videos and completion
      const apiContext = await createApiContext(options, command);
      if (!apiContext) {
        process.exitCode = 1;
        return;
      }

      // Get all courses via API
      const classification = undefined; // all courses
      const courses = await getEnrolledCoursesApi(apiContext.session, { classification });

      apiContext.log.info(`掃描 ${courses.length} 個課程...`);

      // Collect all incomplete videos across all courses using flatMap for cleaner code
      const allIncompleteVideos = (
        await Promise.allSettled(
          courses.map(async (course) => {
            try {
              const videos = await getIncompleteVideosApi(apiContext.session, course.id);
              return videos.map((video) => ({
                courseId: course.id,
                courseName: course.fullname,
                cmid: video.cmid,
                name: video.name,
                url: video.url,
              }));
            } catch (e) {
              apiContext.log.warn(`無法取得課程 ${course.fullname} 的影片: ${e}`);
              return [] as Array<{ courseId: number; courseName: string; cmid: number; name: string; url: string }>;
            }
          })
        )
      )
        .filter((result) => result.status === "fulfilled")
        .flatMap((result) => result.status === "fulfilled" ? result.value : []);

      if (allIncompleteVideos.length === 0) {
        apiContext.log.info("所有影片已完成（或無影片）。");
        return;
      }

      apiContext.log.info(`找到 ${allIncompleteVideos.length} 部未完成影片`);

      // Dry-run: show videos without needing browser
      if (options.dryRun) {
        for (const video of allIncompleteVideos) {
          apiContext.log.info(`  [試執行] [${video.courseName}] ${video.name}`);
        }
        apiContext.log.info("\n===== 執行結果 =====");
        apiContext.log.info(`掃描課程數: ${courses.length}`);
        apiContext.log.info(`找到未完成影片: ${allIncompleteVideos.length}`);
        apiContext.log.info(`執行影片數: ${allIncompleteVideos.length} (試執行)`);
        return;
      }

      // Need browser only for getting viewId and duration (not needed for dry-run)
      const context = await createSessionContext(options, command);
      if (!context) {
        process.exitCode = 1;
        return;
      }

      const { log, page, browser, context: browserContext } = context;

      try {
        const allResults: Array<{ courseName: string; name: string; success: boolean; error?: string }> = [];
        let totalCompleted = 0;
        let totalFailed = 0;

        for (const video of allIncompleteVideos) {
          log.info(`處理中: [${video.courseName}] ${video.name}`);

          try {
            const metadata = await getVideoMetadata(page, video.url, log);

            // Use WS API for completion
            const success = await completeVideoApi(apiContext.session, { ...metadata, cmid: video.cmid.toString() });
            if (success) {
              log.success(`  已完成！`);
              allResults.push({ courseName: video.courseName, name: video.name, success: true });
              totalCompleted++;
            } else {
              log.error(`  失敗。`);
              allResults.push({ courseName: video.courseName, name: video.name, success: false, error: "Failed to complete" });
              totalFailed++;
            }
          } catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            log.error(`  錯誤: ${msg}`);
            allResults.push({ courseName: video.courseName, name: video.name, success: false, error: msg });
            totalFailed++;
          }
        }

        log.info("\n===== 執行結果 =====");
        log.info(`掃描課程數: ${courses.length}`);
        log.info(`找到未完成影片: ${allIncompleteVideos.length}`);
        log.info(`執行影片數: ${totalCompleted}`);
        if (totalFailed > 0) log.warn(`失敗影片數: ${totalFailed}`);

        if (output !== "silent") {
          formatAndOutput(allResults as unknown as Record<string, unknown>[], output, log);
        }
      } finally {
        await closeBrowserSafely(browser, browserContext);
      }
    });

  videosCmd
    .command("download")
    .description("Download videos from a course (requires browser)")
    .argument("<course-id>", "Course ID")
    .option("--output-dir <path>", "Output directory", "./downloads/videos")
    .option("--incomplete-only", "Download only incomplete videos")
    .action(async (courseId, options, command) => {
      const context = await createSessionContext(options, command);
      if (!context) {
        process.exitCode = 1;
        return;
      }

      const { log, page, session, browser, context: browserContext } = context;

      try {
        const videos = await getSupervideosInCourse(page, session, parseInt(courseId, 10), log, {
          incompleteOnly: options.incompleteOnly,
        });

        log.info(`找到 ${videos.length} 個影片`);

        const baseDir = getBaseDir();
        const outputDir = path.resolve(baseDir, options.outputDir);
        fs.mkdirSync(outputDir, { recursive: true });

        const downloaded: Array<{ name: string; path: string; success: boolean; error?: string; type?: string }> = [];

        for (const video of videos) {
          const filename = sanitizeFilename(video.name) + ".mp4";
          const outputPath = path.join(outputDir, filename);

          log.info(`處理中: ${video.name}`);

          try {
            const metadata = await getVideoMetadata(page, video.url, log);
            const result = await downloadVideo(page, metadata, outputPath, log);

            if (result.success) {
              log.success(`  已下載: ${result.path}`);
              downloaded.push({ name: video.name, path: result.path!, success: true, type: result.type });
            } else {
              log.warn(`  失敗: ${result.error}`);
              downloaded.push({ name: video.name, path: "", success: false, error: result.error, type: result.type });
            }
          } catch (err) {
            const msg = err instanceof Error ? err.message : String(err);
            log.error(`  錯誤: ${msg}`);
            downloaded.push({ name: video.name, path: "", success: false, error: msg });
          }
        }

        const completed = downloaded.filter(d => d.success).length;
        const failed = downloaded.filter(d => !d.success).length;
        log.info(`\n執行結果: ${completed} 成功, ${failed} 失敗`);

        console.log(JSON.stringify({
          status: "success",
          timestamp: new Date().toISOString(),
          course_id: courseId,
          output_dir: outputDir,
          total_videos: videos.length,
          downloaded: completed,
          failed,
        }));
        for (const v of downloaded) {
          console.log(JSON.stringify(v));
        }
      } finally {
        await closeBrowserSafely(browser, browserContext);
      }
    });
}

File: src/commands/videos.rs
use anyhow::Result;
use crate::Cli;
use crate::config::load_config;
use crate::moodle::course::get_enrolled_courses_api;
use crate::moodle::video::{get_supervideos_in_course_api, get_incomplete_videos_api, update_completion_status, get_video_metadata_browser, download_video_with_cookies, save_video_progress_api, VideoMetadata};
use crate::output::format_and_output;
use crate::utils::sanitize_filename;
use super::ApiCtx;
use crate::moodle::types::SuperVideoModule;

pub async fn run(cmd: &crate::VideosCommands, cli: &Cli) -> Result<()> {
    let ctx = ApiCtx::build(cli)?;

    match cmd {
        crate::VideosCommands::List { course_id, incomplete_only } => {
            let mut videos = get_supervideos_in_course_api(&ctx.client, &ctx.session, *course_id).await?;
            if *incomplete_only {
                videos.retain(|v| !v.is_complete);
            }

            let items: Vec<serde_json::Value> = videos.iter().map(|v| serde_json::json!({
                "cmid": v.cmid,
                "name": v.name,
                "url": v.url,
                "is_complete": v.is_complete,
            })).collect();

            ctx.log.info(&format!("Found {} videos", items.len()));
            format_and_output(&items, ctx.output, None);
        }

        crate::VideosCommands::Complete { course_id, dry_run } => {
            let videos = get_incomplete_videos_api(&ctx.client, &ctx.session, *course_id).await?;
            let total = videos.len();

            if videos.is_empty() {
                ctx.log.info("All videos already complete (or no videos found).");
                let result = serde_json::json!({
                    "action": "complete",
                    "course_id": *course_id,
                    "total": 0,
                    "completed": 0,
                    "failed": 0,
                    "dry_run": *dry_run,
                });
                format_and_output(&[result], ctx.output, None);
                return Ok(());
            }

            ctx.log.info(&format!("Found {} incomplete videos", total));

            if *dry_run {
                for v in &videos {
                    ctx.log.info(&format!("  [dry-run] {}", v.name));
                }
                ctx.log.info(&format!("Would complete {} videos", total));
                let result = serde_json::json!({
                    "action": "complete",
                    "course_id": *course_id,
                    "total": total,
                    "completed": 0,
                    "failed": 0,
                    "dry_run": true,
                });
                format_and_output(&[result], ctx.output, None);
                return Ok(());
            }

            // Launch browser only for getting metadata
            let config = load_config(cli.config.as_ref().and_then(|p| p.parent()));
            let launched = crate::auth::launch_persistent_session(&config, &ctx.log, true).await?;

            let mut completed = 0;
            for v in &videos {
                ctx.log.info(&format!("Processing: {}", v.name));

                // 1. Get metadata via browser
                let metadata = match get_video_metadata_browser(&launched.page, &v.url, &ctx.log).await {
                    Ok(m) => m,
                    Err(e) => {
                        ctx.log.warn(&format!("  Failed to get metadata: {}", e));
                        continue;
                    }
                };

                // 2. Complete via API
                match complete_video(&ctx, &v, &metadata).await {
                    Ok(()) => {
                        ctx.log.success(&format!("  Completed: {}", v.name));
                        completed += 1;
                    }
                    Err(e) => {
                        ctx.log.warn(&format!("  Failed to complete: {} — {}", v.name, e));
                    }
                }
            }

            crate::auth::close_persistent_session(launched).await;
            ctx.log.info(&format!("Completed {}/{} videos", completed, total));
            let result = serde_json::json!({
                "action": "complete",
                "course_id": *course_id,
                "total": total,
                "completed": completed,
                "failed": total - completed,
                "dry_run": false,
            });
            format_and_output(&[result], ctx.output, None);
        }

        crate::VideosCommands::CompleteAll { dry_run } => {
            let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "inprogress").await?;
            ctx.log.info(&format!("Scanning {} courses for incomplete videos...", courses.len()));

            let mut all_incomplete = Vec::new();
            for course in &courses {
                if let Ok(videos) = get_incomplete_videos_api(&ctx.client, &ctx.session, course.id).await {
                    for v in videos {
                        all_incomplete.push((course.fullname.clone(), v));
                    }
                }
            }

            let total = all_incomplete.len();

            if all_incomplete.is_empty() {
                ctx.log.info("No incomplete videos found.");
                let result = serde_json::json!({
                    "action": "complete_all",
                    "courses_scanned": courses.len(),
                    "total": 0,
                    "completed": 0,
                    "failed": 0,
                    "dry_run": *dry_run,
                });
                format_and_output(&[result], ctx.output, None);
                return Ok(());
            }

            ctx.log.info(&format!("Found {} incomplete videos across all courses", total));

            if *dry_run {
                for (cname, v) in &all_incomplete {
                    ctx.log.info(&format!("  [dry-run] [{}]: {}", cname, v.name));
                }
                ctx.log.info(&format!("Would complete {} videos", total));
                let result = serde_json::json!({
                    "action": "complete_all",
                    "courses_scanned": courses.len(),
                    "total": total,
                    "completed": 0,
                    "failed": 0,
                    "dry_run": true,
                });
                format_and_output(&[result], ctx.output, None);
                return Ok(());
            }

            let config = load_config(cli.config.as_ref().and_then(|p| p.parent()));
            let launched = crate::auth::launch_persistent_session(&config, &ctx.log, true).await?;

            let mut completed = 0;
            for (cname, v) in &all_incomplete {
                ctx.log.info(&format!("Processing [{}]: {}", cname, v.name));

                let metadata = match get_video_metadata_browser(&launched.page, &v.url, &ctx.log).await {
                    Ok(m) => m,
                    Err(e) => {
                        ctx.log.warn(&format!("  Failed to get metadata: {}", e));
                        continue;
                    }
                };

                match complete_video(&ctx, &v, &metadata).await {
                    Ok(()) => {
                        ctx.log.success(&format!("  Completed: {}", v.name));
                        completed += 1;
                    }
                    Err(e) => {
                        ctx.log.warn(&format!("  Failed to complete: {} — {}", v.name, e));
                    }
                }
            }

            crate::auth::close_persistent_session(launched).await;
            ctx.log.info(&format!("Completed {}/{} videos", completed, total));
            let result = serde_json::json!({
                "action": "complete_all",
                "courses_scanned": courses.len(),
                "total": total,
                "completed": completed,
                "failed": total - completed,
                "dry_run": false,
            });
            format_and_output(&[result], ctx.output, None);
        }

        crate::VideosCommands::Download { course_id, output_dir, cmid } => {
            let target_str = cmid.to_string();

            let videos = if let Some(cid) = course_id {
                get_supervideos_in_course_api(&ctx.client, &ctx.session, *cid).await?
            } else {
                let courses = get_enrolled_courses_api(&ctx.client, &ctx.session, "inprogress").await?;
                ctx.log.info(&format!("Scanning {} courses for cmid={}...", courses.len(), cmid));
                let mut found = Vec::new();
                for c in &courses {
                    if let Ok(vs) = get_supervideos_in_course_api(&ctx.client, &ctx.session, c.id).await {
                        for v in vs {
                            if v.cmid == target_str {
                                found.push(v);
                            }
                        }
                    }
                }
                found
            };

            if videos.is_empty() {
                ctx.log.info(&format!("No video found with cmid={}", cmid));
                return Ok(());
            }

            download_videos(&ctx, cli, videos, output_dir).await?;
        }

        crate::VideosCommands::DownloadAll { course_id, output_dir, incomplete_only } => {
            let mut videos = get_supervideos_in_course_api(&ctx.client, &ctx.session, *course_id).await?;
            if *incomplete_only {
                videos.retain(|v| !v.is_complete);
            }

            if videos.is_empty() {
                ctx.log.info("No videos to download.");
                return Ok(());
            }

            download_videos(&ctx, cli, videos, output_dir).await?;
        }
    }

    Ok(())
}

async fn download_videos(ctx: &ApiCtx, cli: &Cli, videos: Vec<SuperVideoModule>, output_dir: &std::path::PathBuf) -> Result<()> {
    ctx.log.info(&format!("Found {} videos", videos.len()));
    tokio::fs::create_dir_all(output_dir).await?;

    let config = load_config(cli.config.as_ref().and_then(|p| p.parent()));
    let launched = crate::auth::launch_persistent_session(&config, &ctx.log, true).await?;

    let cookies = match crate::auth::get_cookies(&launched.page).await {
        Ok(c) => c,
        Err(e) => {
            crate::auth::close_persistent_session(launched).await;
            anyhow::bail!("Failed to extract cookies: {}", e);
        }
    };

    let mut results: Vec<serde_json::Value> = Vec::new();
    let mut downloaded = 0usize;
    let mut failed = 0usize;

    for v in &videos {
        ctx.log.info(&format!("Processing: {}", v.name));

        let metadata = match get_video_metadata_browser(&launched.page, &v.url, &ctx.log).await {
            Ok(m) => m,
            Err(e) => {
                ctx.log.warn(&format!("  Failed to get metadata: {}", e));
                results.push(serde_json::json!({
                    "name": v.name, "success": false, "error": e.to_string(),
                }));
                failed += 1;
                continue;
            }
        };

        let direct_url = metadata.video_sources.iter().find(|s| {
            s.contains("pluginfile.php") || s.ends_with(".mp4") || s.ends_with(".webm")
        });

        if let Some(url) = direct_url {
            let filename = sanitize_filename(&v.name, 200);
            let output_path = output_dir.join(format!("{}.mp4", filename));
            let path_str = output_path.to_string_lossy().to_string();

            match download_video_with_cookies(&cookies, url, &path_str, &ctx.log).await {
                Ok(size) => {
                    ctx.log.success(&format!("  Downloaded: {} ({:.1} KB)",
                        v.name, size as f64 / 1024.0));
                    results.push(serde_json::json!({
                        "name": v.name, "success": true, "path": path_str,
                        "type": "direct", "size": size,
                    }));
                    downloaded += 1;
                }
                Err(e) => {
                    ctx.log.warn(&format!("  Download failed: {}", e));
                    results.push(serde_json::json!({
                        "name": v.name, "success": false, "error": e.to_string(),
                    }));
                    failed += 1;
                }
            }
        } else if !metadata.youtube_ids.is_empty() {
            let yt_url = format!("https://www.youtube.com/watch?v={}", metadata.youtube_ids[0]);
            ctx.log.warn(&format!("  YouTube video: {}", yt_url));
            ctx.log.info("  Use yt-dlp to download YouTube videos.");
            results.push(serde_json::json!({
                "name": v.name, "success": false,
                "error": format!("YouTube video — use yt-dlp: yt-dlp {}", yt_url),
                "type": "youtube",
            }));
            failed += 1;
        } else {
            ctx.log.warn("  No downloadable video source found (embedded/blob URL).");
            results.push(serde_json::json!({
                "name": v.name, "success": false,
                "error": "No downloadable video source found",
                "type": "embedded",
            }));
            failed += 1;
        }
    }

    crate::auth::close_persistent_session(launched).await;

    ctx.log.info(&format!("\nResult: {} downloaded, {} failed", downloaded, failed));
    format_and_output(&results, ctx.output, None);
    Ok(())
}

/// Try to mark a video as complete via supervideo WS API, falling back to completion status.
async fn complete_video(ctx: &ApiCtx, v: &SuperVideoModule, metadata: &VideoMetadata) -> anyhow::Result<()> {
    ctx.log.debug(&format!(
        "  metadata: view_id={:?}, duration={:?}, sources={}, yt_ids={}",
        metadata.view_id, metadata.duration, metadata.video_sources.len(), metadata.youtube_ids.len()
    ));

    if let Some(view_id) = metadata.view_id {
        let duration = metadata.duration;
        let ok = save_video_progress_api(&ctx.client, &ctx.session, view_id, duration)
            .await
            .map_err(|e| anyhow::anyhow!("save_progress (view_id={}): {}", view_id, e))?;
        if !ok {
            anyhow::bail!("save_progress returned success=false (view_id={}, duration={})", view_id, duration);
        }
    } else {
        ctx.log.warn(&format!(
            "  No view_id/duration in page metadata, falling back to completion status API (cmid={})",
            v.cmid
        ));
        let cmid: u64 = v.cmid.parse().map_err(|_| anyhow::anyhow!("invalid cmid: {}", v.cmid))?;
        if cmid == 0 {
            anyhow::bail!("cmid is 0, cannot complete via fallback API");
        }
        let ok = update_completion_status(&ctx.client, &ctx.session, cmid, true)
            .await
            .map_err(|e| anyhow::anyhow!("update_completion (cmid={}): {}", cmid, e))?;
        if !ok {
            anyhow::bail!("update_completion returned status=false (cmid={})", cmid);
        }
    }
    Ok(())
}
